### PR TITLE
AnalyzerDriver async

### PR DESCRIPTION
**Pull request by pharring**
Removed async from ProcessEventAsync and added it to ProcessCompilationUnitCompleted where it was missing.

### CODE REVIEW

**Comment by pharring** [comment]
@mavasani @JohnHamby review please.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 1]
```diff
@@ -490,7 +489,7 @@ private Task ProcessCompilationUnitCompleted(CompilationUnitCompletedEvent compl
                     tasks.Add(task); 
                 }
 
-                return Task.WhenAll(tasks.ToArrayAndFree());
+                await Task.WhenAll(tasks.ToArrayAndFree()).ConfigureAwait(false);
```
Comment (pharring) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: I think this was a bug. There is a finally block and I assume the intent was to execute the 'finally' only after all tasks have completed. Without the await, though, it'll run before the 'return'.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 2]
```diff
@@ -355,33 +355,32 @@ protected AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerM
             }
         }
 
-        private async Task ProcessEventAsync(CompilationEvent e, CancellationToken cancellationToken)
+        private static readonly Task s_emptyTask = Task.FromResult<object>(null);
```
Comment (pharring) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: I couldn't find the equivalent of the Workspace layer's "SpecializedTasks.Empty" in the compiler layer. Is there one?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 3]
```diff
@@ -355,33 +355,32 @@ protected AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerM
...
+
+        private Task ProcessEventAsync(CompilationEvent e, CancellationToken cancellationToken)
```
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Should we drop "Async" from the name?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 4]
```diff
@@ -467,7 +466,7 @@ protected static void AddAnalyzerActionsExecutor(IDictionary<DiagnosticAnalyzer,
 
         protected abstract void AddTasksForExecutingDeclaringReferenceActions(SymbolDeclaredCompilationEvent symbolEvent, IDictionary<DiagnosticAnalyzer, Task> taskMap, CancellationToken cancellationToken);
 
-        private Task ProcessCompilationUnitCompleted(CompilationUnitCompletedEvent completedEvent, CancellationToken cancellationToken)
+        private async Task ProcessCompilationUnitCompleted(CompilationUnitCompletedEvent completedEvent, CancellationToken cancellationToken)
```
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Add "Async" to the name.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 5]
```diff
@@ -511,7 +510,7 @@ private Task ProcessCompilationUnitCompleted(CompilationUnitCompletedEvent compl
             }
         }
 
-        private async Task ExecuteCompilationActionsAsync(ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<CompilationAnalyzerAction>> compilationActionsMap, CancellationToken cancellationToken)
+        private Task ExecuteCompilationActionsAsync(ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<CompilationAnalyzerAction>> compilationActionsMap, CancellationToken cancellationToken)
```
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Drop "Async" from the name?

**Comment by JohnHamby** [comment]
Assuming this change provides a performance benefit, I sign off. @heejaechang and @srivatsn might also take an interest.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 2 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: I don't think compiler layer has one

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 1 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: nice!

**Comment by heejaechang** [comment]
:+1:

**Comment by mavasani** [comment]
:+1: Thanks for cleaning it up!

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 3 above)
Comment (pharring) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: As I understand it, the Async suffix means "task returning" whereas the 'async' (contextual) keyword is a (language) artifact of the implementation.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 4 above)
Comment (pharring) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Agreed.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 5 above)
Comment (sharwell) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: No, the Async suffix corresponds to the return type (and thus the manner in which the method is called), not to the manner in which it is implemented.

**Comment by sharwell** [comment]
:+1:

**Pull request merged**